src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -1347,35 +1346,40 @@ internal static object CompareOperators(ExecutionContext context, IScriptExtent
         }
 
         /// <summary>
-        /// Cache regular expressions...
+        /// Cache regular expressions.
         /// </summary>
         /// <param name="patternString">The string to find the pattern for.</param>
-        /// <param name="options">The options used to create the regex...</param>
-        /// <returns>A case-insensitive Regex...</returns>
+        /// <param name="options">The options used to create the regex.</param>
+        /// <returns>New or cached Regex.</returns>
         internal static Regex NewRegex(string patternString, RegexOptions options)
         {
-            if (options != RegexOptions.IgnoreCase)
-                return new Regex(patternString, options);
-
-            lock (s_regexCache)
+            var subordinateRegexCache = s_regexCache.GetOrAdd(options, s_subordinateRegexCacheCreationDelegate);
+            if (subordinateRegexCache.TryGetValue(patternString, out Regex result))
             {
-                Regex result;
-                if (s_regexCache.TryGetValue(patternString, out result))
-                {
-                    return result;
-                }
-                else
+                return result;
+            }
+            else
+            {
+                return subordinateRegexCache.GetOrAdd(patternString, key =>
                 {
-                    if (s_regexCache.Count > MaxRegexCache)
-                        s_regexCache.Clear();
-                    Regex re = new Regex(patternString, RegexOptions.IgnoreCase);
-                    s_regexCache.Add(patternString, re);
-                    return re;
-                }
+                        if (subordinateRegexCache.Count > MaxRegexCache)
+                        {
+                            // TODO: it would be usefull to get a notice (in telemetry?) if the cache is full.
+                            subordinateRegexCache.Clear();
```

Comment:
Seems overkill to clear the cache.  Would be better if we removed the least used one to make space, but this would mean adding some counting.

Reply:
It looks as overkill only for short time sessions. If we want PowerShell as default shell we should clear/reset/dispose resources/caches otherwise (I mean not only the code) we fall in memory leak and performance degradation with time. 
Also it was in original code too.

Reply:
@SteveL-MSFT Can we continue?

Reply:
@SteveL-MSFT Please update your review.